File: app/src/test/java/com/bank/testeandroidv2/loginScreen/LoginPresenterUnitTest.java
package com.bank.testeandroidv2.loginScreen;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.lang.ref.WeakReference;

@RunWith(RobolectricTestRunner.class)
public class LoginPresenterUnitTest {
    public static String TAG = LoginPresenterUnitTest.class.getSimpleName();

    @Before
    public void setUp() {

    }

    @After
    public void tearDown() {

    }

    @Test
    public void presentLoginDataValidationForm_with_validInput_shouldCall_displayLoginData() {
        //Given
        LoginPresenter loginPresenter = new LoginPresenter();
        LoginResponse loginResponse = new LoginResponse();

        LoginActivityInputSpy loginActivityInputSpy = new LoginActivityInputSpy();
        loginPresenter.output = new WeakReference<LoginActivityInput>(loginActivityInputSpy);

        //When
        loginPresenter.presentLoginDataValidationForm(loginResponse);

        //Then
        Assert.assertTrue("When the valid input is passed to LoginPresenter Then displayLoginMetaData should be called",
                loginActivityInputSpy.isProcessLoginDataFormCalled);
    }


    @Test
    public void processRequestLoginDataOnServer_with_success_shouldCall_callNextScene() {
        //Given
        LoginPresenter loginPresenter = new LoginPresenter();
        LoginResponse loginResponse = new LoginResponse();

        LoginActivityInputSpy loginActivityInputSpy = new LoginActivityInputSpy();
        loginPresenter.output = new WeakReference<LoginActivityInput>(loginActivityInputSpy);

        UserAccount uc = new UserAccount();
        uc.userId = 1L;
        uc.name = "Jose da Silva Teste";
        uc.agency = "012314564";
        uc.bankAccount = "2050";
        uc.balance = 3.3445;

        //When
        loginPresenter.processRequestLoginDataOnServer(uc);

        //Then
        Assert.assertTrue(loginActivityInputSpy.isCallNextSceneCalled);
        Assert.assertNotNull(loginActivityInputSpy.userAccountCopy);
    }

    @Test
    public void processErrorRequest_with_error_shouldCall_callApiError() {
        //Given
        LoginPresenter loginPresenter = new LoginPresenter();
        LoginResponse loginResponse = new LoginResponse();

        LoginActivityInputSpy loginActivityInputSpy = new LoginActivityInputSpy();
        loginPresenter.output = new WeakReference<LoginActivityInput>(loginActivityInputSpy);

        //When
        loginPresenter.processErrorRequest(null);

        //Then
        Assert.assertTrue(loginActivityInputSpy.isCallApiErrorCalled);
    }


    private class LoginActivityInputSpy implements LoginActivityInput {
        public boolean isProcessLoginDataFormCalled = false;
        public boolean isCallNextSceneCalled = false;
        public boolean isCallApiErrorCalled = false;
        public LoginViewModel loginViewModelCopy;
        public UserAccount userAccountCopy;
        public String errorCopy;

        @Override
        public void processLoginDataForm(LoginViewModel loginViewModel) {
            isProcessLoginDataFormCalled = true;
            loginViewModelCopy = loginViewModel;
        }

        @Override
        public void callNextScene(UserAccount response) {
            isCallNextSceneCalled = true;
            userAccountCopy = response;
        }

        @Override
        public void callApiError(String error) {
            isCallApiErrorCalled = true;
            errorCopy = error;
        }
    }
}

File: app/src/test/java/br/com/cauejannini/testesantander/LoginPresenterUnitTest.kt
package br.com.cauejannini.testesantander

import android.os.Build
import br.com.cauejannini.testesantander.login.*
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.lang.ref.WeakReference

@Config(sdk = [Build.VERSION_CODES.O_MR1])
@RunWith(RobolectricTestRunner::class)
class LoginPresenterUnitTest {

    @Test
    fun present_correctData_onLoginSuccessfull() {

        val loginPresenter = LoginPresenter()
        val outputSpy = LoginActivityInputSpy()
        loginPresenter.output = WeakReference(outputSpy)

        val loginResponseModel = LoginResponseModel()

        val userAccount = UserAccount()
        userAccount.userId = 1
        userAccount.name = "Lucio dos Santos Teste"
        userAccount.agency = "1234"
        userAccount.bankAccount = "567890"
        userAccount.balance = 990.34

        loginResponseModel.userAccount = userAccount
        loginResponseModel.error = null

        loginPresenter.onLoginResponse(loginResponseModel)

        Assert.assertTrue(outputSpy.onLoggedInCalled)
        Assert.assertEquals(outputSpy.userAccountCopy, userAccount)

        Assert.assertFalse(outputSpy.onLoginErrorCalled)
        Assert.assertNull(outputSpy.errorMessageCopy)

    }

    @Test
    fun present_correctData_onLoginResponseError() {

        val loginPresenter = LoginPresenter()
        val outputSpy = LoginActivityInputSpy()
        loginPresenter.output = WeakReference(outputSpy)

        val loginResponseModel = LoginResponseModel()

        loginResponseModel.userAccount = null

        val errorMessage = "Erro retornado no LoginResponseModel"
        loginResponseModel.error = errorMessage

        loginPresenter.onLoginResponse(loginResponseModel)

        Assert.assertTrue(outputSpy.onLoginErrorCalled)
        Assert.assertEquals(outputSpy.errorMessageCopy, errorMessage)

        Assert.assertFalse(outputSpy.onLoggedInCalled)
        Assert.assertNull(outputSpy.userAccountCopy)

    }

    @Test
    fun present_correctData_onLoginFailed() {

        val loginPresenter = LoginPresenter()
        val outputSpy = LoginActivityInputSpy()
        loginPresenter.output = WeakReference(outputSpy)

        val errorMessage = "Erro de integração"

        loginPresenter.onLoginFailed(errorMessage)

        Assert.assertTrue(outputSpy.onLoginErrorCalled)
        Assert.assertEquals(outputSpy.errorMessageCopy, errorMessage)

        Assert.assertFalse(outputSpy.onLoggedInCalled)
        Assert.assertNull(outputSpy.userAccountCopy)

    }

    private class LoginActivityInputSpy: LoginActivityInput {

        var onLoggedInCalled = false
        var userAccountCopy: UserAccount? = null

        var onLoginErrorCalled = false
        var errorMessageCopy: String? = null

        override fun onLoggedIn(userAccount: UserAccount) {
            onLoggedInCalled = true
            userAccountCopy = userAccount
        }

        override fun onLoginError(message: String) {
            onLoginErrorCalled = true
            errorMessageCopy = message
        }

    }

}